### ISSUE TITLE
"Statement is unreachable" false positive when calling `callable` on union with `object`

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#11764]
```py
def foo(value: object | str) -> None:
    if callable(value):
        print(1) # Statement is unreachable  [unreachable]
```
https://mypy-play.net/?mypy=master&python=3.10&flags=show-error-codes%2Callow-redefinition%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-optional%2Cno-implicit-reexport%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=a72eb2ed08576c38916daccddd89fed7

**Comment by DetachHead** [comment]
oops didn't mean to click feature but i can't remove it

**Comment by sobolevn** [comment]
Sorry, but this is true. `object` is not callable:

```python
>>> o = object()
>>> o()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'object' object is not callable
```

But, `Type[object]` is.

Or am I missing something?

**Comment by DetachHead** [comment]
sure `object` isn't, but a subtype of `object` could be

```py
def foo(value: object | str) -> None:
    if callable(value):
        print(1) # Statement is unreachable  [unreachable]
        
foo(lambda: ...) # no error because a `Callable` is a subtype of `object`

callable(lambda: ...) # True
```

**Comment by sobolevn** [comment]
Subtype of any type can be callble 🙂 

Do you mean that `object` should be treated specially, because it is a base type?

**Comment by DetachHead** [comment]
no it shouldn't be treated specifically, instead `callable` should narrow it. for example see how it's done in typescript

```ts
function foo(value: unknown | string) { //unknown is the typescript equivalent of object
    if (typeof value === 'function') {
        value() //narrowed to Function
    }
}
```
https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUA3AhgGxAUwC5FwBrMOAdyQB9EBnKAJxjAHMBKRAbwFgAoREMQxgiNFACeAB3xwx2PPkQBeNYgDkoSLAQaufQcOOKCaLgHoLYLEyZV8AE0RQ4iAGLho8MAOMBfAX8gA

if you look at the signature of `callable`, that's actually what it should be doing, since it uses a `TypeGuard`:
```py
def callable(__obj: object) -> TypeGuard[Callable[..., object]]: ...
```

in fact, if we just copy that signature and make our own `callable`, the issue goes away:
```py
from typing import TypeGuard, Callable

def callable2(__obj: object) -> TypeGuard[Callable[..., object]]: ...

def foo(value: object | str) -> None:
    if callable2(value):
        reveal_type(value) # note: Revealed type is "def (*Any, **Any) -> builtins.object"
        print(1)

def bar(value: object | str) -> None:
    if callable(value):
        reveal_type(value) # error: Statement is unreachable  [unreachable]
```
https://mypy-play.net/?mypy=latest&python=3.10&flags=warn-unreachable&gist=cc5b40496ef4e540a6fc7789332269ab

so it seems that the problem is that there's some sort of intrinsic behavior on `callable` that should just be removed?

**Comment by sobolevn** [comment]
Oh, now I can see a problem:

```python
from typing import TypeGuard, Callable

def callable2(__obj: object) -> TypeGuard[Callable[..., object]]: ...

def foo(value: object | str) -> None:
    if callable2(value):
        reveal_type(value) # note: Revealed type is "def (*Any, **Any) -> builtins.object"
        print(1)

def bar(value: object | str) -> None:
    if callable(value):
        reveal_type(value) # error: Statement is unreachable  [unreachable]
```

This is inconsustent and probably should be solved. I will take a look, thanks a lot for your explanation! 👍

**Comment by sobolevn** [comment]
Right now in typeshed `callable` is defined as `def callable(__obj: object) -> TypeGuard[Callable[..., object]]: ...`.

But, mypy does some in-depth analysis of `callable()` function. During this process we do partition by callable / non callable types to narrow types more precisely. Source: https://github.com/python/mypy/blob/56684e43a14e3782409c47e99bb47191d631a3de/mypy/checker.py#L4062-L4142

I am not sure how to solve this problem 🤔 

On one hand, I understand that some subtypes of any types can be callble. And `TypeGuard` case shows us this. On the other hand, using types that do not declare `__call__` as their contract and still use them is not very good. I would probably what to see `object | str | SomeCallableProtocol` as the argument type here.

**Comment by KotlinIsland** [comment]
Could we just simplify the union here?
